src/config.c

```diff
@@ -754,6 +754,103 @@ static int cgroup_config_mount_fs(void)
 	return error;
 }
 
+static int convert_controller_versions(struct cgroup *cgroup)
+{
+	enum cg_version_t current_version, convert_version;
+	struct cgroup *cgrp_cpy, *convert_cgrp = NULL;
+	struct cgroup_controller *cgc;
+	int i, j, ret;
+
+	if (!cgroup)
+		return ECGINVAL;
+
+	/* No controllers added */
+	if (cgroup->index == 0)
+		return ECGINVAL;
+
+	cgrp_cpy = cgroup_new_cgroup(cgroup->name);
+	if (!cgrp_cpy) {
+		cgroup_err("Failed to create cgroup %s\n", cgroup->name);
+		return ECGROUPNOTCREATED;
+	}
+
+	/* Copy the non-controllers details */
+	cgrp_cpy->index = 0;
+
+	ret = cgroup_set_uid_gid(cgrp_cpy,
+				 cgroup->tasks_uid, cgroup->tasks_gid,
+				 cgroup->control_uid, cgroup->control_gid);
+	/* cgroup existence is checked, so ignore ret value. */
+
+	cgroup_set_permissions(cgrp_cpy, cgroup->control_dperm,
+			       cgroup->control_fperm, cgroup->task_fperm);
+
+	for (i = 0; i < cgroup->index; i++) {
+		current_version = cgroup->controller[i]->version;
+		convert_version = current_version == CGROUP_V1 ? CGROUP_V2 : CGROUP_V1;
+
+		cgc = cgroup_add_controller(cgrp_cpy, cgroup->controller[i]->name);
+		if (!cgc) {
+			cgroup_err("Failed to allocated controller.\n");
+			ret = ECGFAIL;
+			goto err;
+		}
+
+		for (j = 0; j < cgroup->controller[i]->index; j++) {
+			ret = cgroup_add_value_string(cgc,
+						      cgroup->controller[i]->values[j]->name,
+						      cgroup->controller[i]->values[j]->value);
+			if (ret) {
+				cgroup_err("Failed to add setting to %s\n", cgc->name);
+				goto err;
+			}
+		}
+
+		/*
+		 * There might be more than one controller parsed from the
+		 * cgconfig.conf file, let's create cgroup with one controller
+		 * and see if it fails and if it does, let's try converting
+		 * it and retry the cgroup creation, otherwise move on to
+		 * the next controller.
+		 */
+		ret = cgroup_create_cgroup(cgrp_cpy, 0);
+		if (!ret)
+			goto next_controller;
+
+		/*
+		 * The controller version != cgroup->controller[i] version,
+		 * let try to convert the controller to the current version
+		 * and retry.
+		 */
+		convert_cgrp = cgroup_new_cgroup(cgroup->name);
+		if (!convert_cgrp) {
+			cgroup_err("Failed to created cgroup %s\n", cgroup->name);
+			ret = ECGROUPNOTCREATED;
+			goto err;
+		}
+		ret = cgroup_convert_cgroup(convert_cgrp, current_version, cgrp_cpy,
```

Comment:
Help me out here.  I think this is a "me" problem :)

Here's the simplified prototype for convert:
```cgroup_convert_cgroup(out_cg, out_version, in_cg, in_version)```

So doesn't that mean that we have the versions backward on line 831?  I see that the test passed, so it seems to be working.  But I don't understand how.

Reply:
Great catch, the issue is with the wrong variables used.
```
diff --git a/src/config.c b/src/config.c
index 8d59aca..c5c8bf9 100644
--- a/src/config.c
+++ b/src/config.c
@@ -786,8 +786,8 @@ static int convert_controller_versions(struct cgroup *cgroup)
                               cgroup->control_fperm, cgroup->task_fperm);
 
        for (i = 0; i < cgroup->index; i++) {
-               current_version = cgroup->controller[i]->version;
-               convert_version = current_version == CGROUP_V1 ? CGROUP_V2 : CGROUP_V1;
+               convert_version = cgroup->controller[i]->version;
+               current_version = convert_version == CGROUP_V1 ? CGROUP_V2 : CGROUP_V1;
 
                cgc = cgroup_add_controller(cgrp_cpy, cgroup->controller[i]->name);
                if (!cgc) {
@@ -828,8 +828,8 @@ static int convert_controller_versions(struct cgroup *cgroup)
                        ret = ECGROUPNOTCREATED;
                        goto err;
                }
-               ret = cgroup_convert_cgroup(convert_cgrp, current_version, cgrp_cpy,
-                                           convert_version);
+               ret = cgroup_convert_cgroup(convert_cgrp, convert_version, cgrp_cpy,
+                                           current_version);
                if (ret) {
                        cgroup_err("Conversion of controller failed.\n");
```
will respin v4 with the comments `current_version` and `convert_version`